pkg/extension/hook.go

```diff
@@ -0,0 +1,160 @@
+package extension
+
+import (
+	"context"
+	"sort"
+
+	"github.com/samber/lo"
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/trivy/pkg/flag"
+	"github.com/aquasecurity/trivy/pkg/types"
+)
+
+var hooks = make(map[string]Hook)
+
+func RegisterHook(s Hook) {
+	// Avoid duplication
+	hooks[s.Name()] = s
+}
+
+func DeregisterHook(name string) {
+	delete(hooks, name)
+}
+
+// Hook is an interface that defines the methods for a hook.
+type Hook interface {
+	// Name returns the name of the extension.
+	Name() string
+}
+
+// RunHook is a extension that is called before and after all the processes.
+type RunHook interface {
+	Hook
+
+	// PreRun is called before all the processes.
+	PreRun(ctx context.Context, opts flag.Options) error
+
+	// PostRun is called after all the processes.
+	PostRun(ctx context.Context, opts flag.Options) error
+}
+
+// ScanHook is a extension that is called before and after the scan.
+type ScanHook interface {
+	Hook
+
+	// PreScan is called before the scan. It can modify the scan target.
+	// It may be called on the server side in client/server mode.
+	PreScan(ctx context.Context, target *types.ScanTarget, opts types.ScanOptions) error
+
+	// PostScan is called after the scan. It can modify the results.
+	// It may be called on the server side in client/server mode.
+	// NOTE: Wasm modules cannot directly modify the passed results,
+	//       so it returns a copy of the results.
+	PostScan(ctx context.Context, results types.Results) (types.Results, error)
+}
+
+// ReportHook is a extension that is called before and after the report is written.
+type ReportHook interface {
+	Hook
+
+	// PreReport is called before the report is written.
+	// It can modify the report. It is called on the client side.
+	PreReport(ctx context.Context, report *types.Report, opts flag.Options) error
+
+	// PostReport is called after the report is written.
+	// It can modify the report. It is called on the client side.
+	PostReport(ctx context.Context, report *types.Report, opts flag.Options) error
+}
+
+func PreRun(ctx context.Context, opts flag.Options) error {
+	for _, e := range hooks {
```

Comment:
We will check hooks in random order.
So there may be a conflict when 2 hooks update the same result/value/etc.

I am not suggesting to change anything, but since we are not updating docs - let's add a comment about it.

Reply:
It's better that the order is cosnstent. Updated.
https://github.com/aquasecurity/trivy/pull/8585/commits/401be5e4446713c5cd08bcc3460acb73576c9d6c